### ISSUE TITLE
RGB Led Strip Control

### ISSUE THREAD

**Issue by TimTower01** [mariusmotea/diyHue#63]
I got everything working with the neopixels. But i am not able to get it to work with my RGB 5050 Led Strip.

My current setup looks like this:
![untitled sketch_steckplatine](https://user-images.githubusercontent.com/31222710/31356561-c2646fda-ad3e-11e7-9cd4-3f2085fbd1eb.png)
But my leds are only glow a little bit and dont let me control them.
Any help is desired.

~Tim

**Comment by mariusmotea** [comment]
Hi,

If you connect for short time the base of NPN transistor to +3.3v it will turn that colour to maximum brightness? I believe ESP8622 don't provide enough current for your npn transistors, better to convert them to Darlington with another 3 small extra transistors. The best will be to use N-Channel mosfe transistor that work better with pwm signal.

Marius.

**Comment by cad435** [comment]
Hey, just stepped by,

the reason why its not proberly working is that you don't have a Resistor between the Base of your NPN and the driving pin. I'd suggest something between 1k and 10k, usually I use 2.2k

so, keep in mind, Bipolar Junktion transistors are CURRENT driven CURRENT regulators, so in fact the current you can drive calculates like: I_Driven = I_base x h_FE (where 'h_FE' is sort of an "amplifying factor" you can find in the datasheet)
So for excample:
with a 2.2K resistor before the basse you have a current flow of I = U/R --> I = 3.3V/2200Ohm = 1.5mA
with a BC547C (http://www.ece.usu.edu/ece_store/spec/BC547.pdf) and a h_FE of lets say 500 you can drive 1.5mA x 500 = 750mA (wich is about 1-2m LED strip)


so if you want more LEDs you need a Transistor with a greater h_FE (a darlington transistor for excample) or a smaller base resistor (if the ESP can drive more current!)

OR you switch to mosfets, wich for this task have only benefits compared to the Bipolar Junktion transistors (="BJT") (wich are commonly named as "transistors")
--> fewer heat dissipation compared to BJT
--> faster switching
--> lesser load on the microcontroller (MOSFETS don't ned a flowing current trough its gate)
--> MUCH more current (in fact a 0.5x0.5cm big MOSFET can drive 10A and more)

I'll attach a picture here of such an "Amplifier" https://teachmetomake.files.wordpress.com/2011/05/fetswitch.jpg

the only thing, you need a MOSFET wich can operate at 3.3V gate voltage, usually I take "IRLU024" MOSFETS wich can drive 10A at 3.3V Gate Voltage

**Comment by TimTower01** [comment]
Ok i use a Powermosfet called IRFZ44N. 
The Link for the datasheet: [http://www.produktinfo.conrad.com/datenblaetter/150000-174999/162747-da-01-en-IRFZ_44_N.pdf](url)
Cant figure out if they are suitable for the esp because on the arduino uno they work like a charm.
Maybe you can help me

Tim

**Comment by cad435** [comment]
Hey,

As you can see in figure 3 (Page 3) at the datasheet, the IRFZ44N can drive 10A @ 5V gate voltage.
So for your UNO it'll work well, because your UNO have 5V pins.


as for the ESP8266 with 3.3V gate voltage it'll will not work, because as you might notice in figure 3 again the graph stops at 4.5V.

So 3.3V are not high enough to "open" the gate and let the currrent flow trough.

you can use a PNP (BJT) transistor to amplify the output of the ESP-pin.
I may draw a circuit but right now I have no time for it, sry.

cad435

**Comment by TimTower01** [comment]
Can i use a 3.3V to 5V levelshifter to do this because i have on spare at the moment ?

**Comment by mariusmotea** [comment]
This may not be the most simple, but it will work very good:
https://goo.gl/images/bzRnw1

Marius.

**Comment by cad435** [comment]
@Tim: yes i guess you can use levelshifter but they have to be fast enough (Marius use 5kHz for the PWM, so you need a shifter faster than 5kHz)

cad435

**Comment by erazor666** [comment]
@cad435 
I'm using IRLB871 mosfets for this application, will this 20% loss apply for this mosfet aswell (same as the IRFZ44N)?

**Comment by cad435** [comment]
@erazor666
Hey,

i cant find a Datasheet for your named MOSFET, so i can't tell if its working or not...

And were did you put the 1k resistor?
In series to the LED strips? That'll make no sense....

**Comment by erazor666** [comment]
![photo_2017-11-10_03-27-16](https://user-images.githubusercontent.com/31832864/33435016-5e8e7610-d5e1-11e7-8bdd-012dd23ceb12.jpg)

This is a RGB strip controller i built, the resistors in the picture are 1k.
These are the mosfets: https://www.aliexpress.com/item/FREE-SHIPPING-IRLB8721-RLB8721-LB8721-B8721-8721-TO220-20PCS/32762456412.html

They do work, and i dont have any issues with them, just woundering if they are giving the full potential of the led strip. A pwm dimmer for a single color or white strip will use only one of these mosfets of course, the picture is for illustration only.

**Comment by cad435** [comment]
Alright then, 

so, first things first:
You have a MOSFET, not a NPN-Bipolar Junction Transistor, so you generally DONT'T have a Voltage drop of 0,7V and therefore don't have the 20% Light-Loss

I can't get the Idea of your circuit, but you write its working, so I assume you've built it right (see the circuit Timtower01 postet above, thats totally fine, but you can leave the 1k Resistor at the Gate of the mosfet, that will make no difference. It's infact a bit of Protection for the microcotroller, so its all OK :)

I have the datasheet of the IRLB8721 here:
https://cdn-shop.adafruit.com/datasheets/irlb8721pbf.pdf

so as you can see in Figure 3 at 3.3V Gate this mosfet is capable of about 3-4A wich should be good as long as your LED strip is not longer than 5m.

So, TL;DR:
Read the datasheet, your MOSFET is perfectly fine :)

**Comment by erazor666** [comment]
Hi again and thanks for you reply.
I haven't been making electronics for very long, and the circuit was kinda pieced together from things i found online and a bit of forum reading here and there. Not sure its optimal/correct other than it works. From my limited logic sense of things i couldnt really see what the resistors towards the common ground for each MOSFET are really "doing".

It may be confusing for the backside of the diagram is actully drawn wrong, if you had x-ray vision and looked through the board its correct :) you seemed to have some quite advanced knowledge about this so i thought it was a good place to ask.

**Comment by cad435** [comment]
ah, ok, now i see, didn't know thats the Back-side of your PCB^^

the resistors at the gates are just pulldowns, so if you power the strips and the microcontroller is not connected or is defect, the MOSFET is automatically turned off and the LEDs won't light up uncontrolled...

but as the ESP8266 has a very limited current range on its pins (+-6mA) I would suggest you take a somewhat higher resistance (I usually use 2.2k) but if you have already built your circuit it into some case or something, just leave the resistors, they'll work fine...

**Comment by erazor666** [comment]
Its a prototype mostly, so for the next one i'll just swap to 2.2k ones.  Still optimizing the design, trying to make them as small and robust as possible as i intend to use this for a lot of lights around the house. When its as small and optimal as i can make i'll get some 3d printed boxes or such to house them.
I'm using female pin headers where ever i can so components can be removable in case something breaks.

**Comment by vKnmnn** [comment]
check out this repo for a complete ESP based board to control various kinds of LED strips: https://github.com/EspLight/
